File: ai/select-algorithm-dotnet/src/CompareAll.cs
/// Unified comparison runner for all 9 combinations (3 algorithms × 3 similarity metrics).
/// Executes vector searches sequentially for fair timing and prints a formatted comparison table.

namespace SelectAlgorithm;

using System.Diagnostics;
using MongoDB.Driver;
using MongoDB.Bson;
using OpenAI.Embeddings;

public static class CompareAll
{
    private record IndexConfig(string Name, string Kind, string Similarity, BsonDocument ExtraParams);

    private record SearchResult(string IndexName, string Algorithm, string Metric, long LatencyMs, List<BsonDocument> Results);

    public static void Run()
    {
        Console.WriteLine(new string('=', 60));
        Console.WriteLine("  Compare All Algorithms × Metrics");
        Console.WriteLine("  9 combinations: IVF, HNSW, DiskANN × COS, L2, IP");
        Console.WriteLine(new string('=', 60));

        var databaseName = Environment.GetEnvironmentVariable("AZURE_DOCUMENTDB_DATABASENAME") ?? "Hotels";
        var dataFile = Environment.GetEnvironmentVariable("DATA_FILE_WITH_VECTORS") ?? "../../data/Hotels_Vector.json";
        var vectorField = Environment.GetEnvironmentVariable("EMBEDDED_FIELD") ?? "DescriptionVector";
        var dimensions = int.Parse(Environment.GetEnvironmentVariable("EMBEDDING_DIMENSIONS") ?? "1536");
        var batchSize = int.Parse(Environment.GetEnvironmentVariable("LOAD_SIZE_BATCH") ?? "100");
        var queryText = Environment.GetEnvironmentVariable("QUERY_TEXT") ?? "luxury hotel near the beach";
        var topK = int.Parse(Environment.GetEnvironmentVariable("TOP_K") ?? "3");
        var verbose = (Environment.GetEnvironmentVariable("VERBOSE") ?? "false").Equals("true", StringComparison.OrdinalIgnoreCase);

        var mongoClient = Utils.GetMongoClientPasswordless();
        var embeddingClient = Utils.GetEmbeddingClient();

        try
        {
            var database = mongoClient.GetDatabase(databaseName);

            // Drop collection if it already exists (clean start)
            var collectionNames = database.ListCollectionNames().ToList();
            if (collectionNames.Contains("hotels"))
            {
                database.DropCollection("hotels");
                Console.WriteLine("Dropped existing 'hotels' collection.");
            }

            var collection = database.GetCollection<BsonDocument>("hotels");

            // Load data once into single collection
            var data = Utils.ReadJsonFile(dataFile);
            var documents = data.Where(d => d.Contains(vectorField)).ToList();
            Console.WriteLine($"\nLoaded {documents.Count} documents with embeddings");
            Utils.InsertData(collection, documents, batchSize);

            // Generate ONE embedding for the query (reused for all 9 searches)
            Console.WriteLine($"\nQuery: \"{queryText}\"");
            Console.WriteLine($"Top K: {topK}");
            var embeddingResult = embeddingClient.GenerateEmbedding(queryText);
            var queryVector = embeddingResult.Value.ToFloats().ToArray();
            Console.WriteLine("Embedding generated (reused for all searches)\n");

            // Define 9 index configurations
            var configs = BuildIndexConfigs(dimensions);

            // Create all 9 indexes (idempotent)
            Console.WriteLine("Creating 9 vector indexes...");
            foreach (var config in configs)
            {
                CreateIndex(collection, vectorField, config);
            }
            Console.WriteLine("Waiting for indexes to build...");
            Thread.Sleep(5000);

            // Run searches sequentially for fair timing
            Console.WriteLine("\nRunning searches...\n");
            var results = new List<SearchResult>();
            foreach (var config in configs)
            {
                var sw = Stopwatch.StartNew();
                var searchResults = RunVectorSearch(collection, queryVector, vectorField, config.Name, topK);
                sw.Stop();

                results.Add(new SearchResult(config.Name, config.Kind, config.Similarity, sw.ElapsedMilliseconds, searchResults));

                if (verbose)
                {
                    Console.WriteLine($"  {config.Name}: {sw.ElapsedMilliseconds}ms ({searchResults.Count} results)");
                }
            }

            // Print comparison table
            PrintComparisonTable(results, verbose);
        }
        finally
        {
            // Cleanup: drop the comparison collection
            try
            {
                var database = mongoClient.GetDatabase(databaseName);
                database.DropCollection("hotels");
                Console.WriteLine("\nCleanup: dropped collection 'hotels'");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Cleanup warning: {ex.Message}");
            }
            mongoClient.Cluster.Dispose();
        }
    }

    private static List<IndexConfig> BuildIndexConfigs(int dimensions)
    {
        string[] metrics = ["COS", "L2", "IP"];
        var configs = new List<IndexConfig>();

        foreach (var metric in metrics)
        {
            configs.Add(new IndexConfig(
                $"vector_ivf_{metric.ToLower()}",
                "vector-ivf",
                metric,
                new BsonDocument { { "numLists", 1 } }
            ));

            configs.Add(new IndexConfig(
                $"vector_hnsw_{metric.ToLower()}",
                "vector-hnsw",
                metric,
                new BsonDocument { { "m", 16 }, { "efConstruction", 64 } }
            ));

            configs.Add(new IndexConfig(
                $"vector_diskann_{metric.ToLower()}",
                "vector-diskann",
                metric,
                new BsonDocument { { "maxDegree", 32 }, { "lBuild", 50 } }
            ));
        }

        return configs;
    }

    private static void CreateIndex(IMongoCollection<BsonDocument> collection, string vectorField, IndexConfig config)
    {
        // Drop existing index with same name if present
        try
        {
            collection.Indexes.DropOne(config.Name);
        }
        catch (MongoCommandException)
        {
            // Index doesn't exist, that's fine
        }

        var cosmosSearchOptions = new BsonDocument
        {
            { "kind", config.Kind },
            { "dimensions", int.Parse(Environment.GetEnvironmentVariable("EMBEDDING_DIMENSIONS") ?? "1536") },
            { "similarity", config.Similarity }
        };

        foreach (var param in config.ExtraParams)
        {
            cosmosSearchOptions.Add(param);
        }

        var command = new BsonDocument
        {
            { "createIndexes", collection.CollectionNamespace.CollectionName },
            { "indexes", new BsonArray
                {
                    new BsonDocument
                    {
                        { "name", config.Name },
                        { "key", new BsonDocument(vectorField, "cosmosSearch") },
                        { "cosmosSearchOptions", cosmosSearchOptions }
                    }
                }
            }
        };

        try
        {
            collection.Database.RunCommand<BsonDocument>(command);
        }
        catch (MongoCommandException ex) when (ex.Message.Contains("already exists"))
        {
            // Index already exists with same config — idempotent
        }
    }

    private static List<BsonDocument> RunVectorSearch(
        IMongoCollection<BsonDocument> collection,
        float[] queryVector,
        string vectorField,
        string indexName,
        int topK)
    {
        var pipeline = new[]
        {
            new BsonDocument("$search", new BsonDocument("cosmosSearch", new BsonDocument
            {
                { "vector", new BsonArray(queryVector.Select(f => (double)f)) },
                { "path", vectorField },
                { "k", topK }
            })),
            new BsonDocument("$project", new BsonDocument
            {
                { "HotelName", 1 },
                { "score", new BsonDocument("$meta", "searchScore") }
            })
        };

        return collection.Aggregate<BsonDocument>(pipeline).ToList();
    }

    private static void PrintComparisonTable(List<SearchResult> results, bool verbose)
    {
        Console.WriteLine();
        Console.WriteLine(new string('=', 78));
        Console.WriteLine("  COMPARISON RESULTS");
        Console.WriteLine(new string('=', 78));
        Console.WriteLine();

        // Header
        var header = "Index Name".PadRight(24) +
                     "Algorithm".PadRight(14) +
                     "Metric".PadRight(8) +
                     "Latency".PadRight(10) +
                     "Top Result".PadRight(22);
        Console.WriteLine(header);
        Console.WriteLine(new string('-', 78));

        foreach (var result in results)
        {
            var topResult = "—";
            var topScore = "";
            if (result.Results.Count > 0)
            {
                var doc = result.Results[0];
                topResult = doc.Contains("HotelName") ? doc["HotelName"].AsString : "Unknown";
                if (topResult.Length > 18) topResult = topResult[..18] + "...";
                var score = doc.Contains("score") ? doc["score"].ToDouble() : 0.0;
                topScore = $" ({score:F3})";
            }

            var algoDisplay = result.Algorithm.Replace("vector-", "").ToUpper();
            var row = result.IndexName.PadRight(24) +
                      algoDisplay.PadRight(14) +
                      result.Metric.PadRight(8) +
                      $"{result.LatencyMs}ms".PadRight(10) +
                      $"{topResult}{topScore}";
            Console.WriteLine(row);
        }

        Console.WriteLine(new string('-', 78));
        Console.WriteLine();

        // Summary stats
        var fastest = results.MinBy(r => r.LatencyMs)!;
        var slowest = results.MaxBy(r => r.LatencyMs)!;
        Console.WriteLine($"  Fastest: {fastest.IndexName} ({fastest.LatencyMs}ms)");
        Console.WriteLine($"  Slowest: {slowest.IndexName} ({slowest.LatencyMs}ms)");
        Console.WriteLine();

        if (verbose)
        {
            Console.WriteLine("  DETAILED RESULTS:");
            Console.WriteLine();
            foreach (var result in results)
            {
                Console.WriteLine($"  [{result.IndexName}]");
                for (var i = 0; i < result.Results.Count; i++)
                {
                    var doc = result.Results[i];
                    var name = doc.Contains("HotelName") ? doc["HotelName"].AsString : "Unknown";
                    var score = doc.Contains("score") ? doc["score"].ToDouble() : 0.0;
                    Console.WriteLine($"    {i + 1}. {name} (score: {score:F4})");
                }
                Console.WriteLine();
            }
        }
    }
}

File: ai/select-algorithm-java/src/main/java/com/azure/documentdb/selectalgorithm/CompareAll.java
package com.azure.documentdb.selectalgorithm;

import com.azure.ai.openai.OpenAIClient;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoDatabase;
import org.bson.Document;

import java.util.ArrayList;
import java.util.List;

/**
 * Unified comparison runner that executes all 9 combinations
 * (3 algorithms x 3 similarity metrics) and prints a formatted table.
 */
public class CompareAll {

    private static final String COLLECTION_NAME = "hotels";
    private static final String[] ALGORITHMS = {"ivf", "hnsw", "diskann"};
    private static final String[] METRICS = {"COS", "L2", "IP"};

    public static void main(String[] args) {
        run();
    }

    public static void run() {
        String queryText = Utils.getEnv("QUERY_TEXT", "luxury hotel near the beach");
        int topK = Integer.parseInt(Utils.getEnv("TOP_K", "5"));

        String databaseName = Utils.getEnv("AZURE_DOCUMENTDB_DATABASENAME", "Hotels");
        String dataFile = Utils.getEnv("DATA_FILE_WITH_VECTORS", "data/Hotels_Vector.json");
        String vectorField = Utils.getEnv("EMBEDDED_FIELD", "DescriptionVector");
        int dimensions = Integer.parseInt(Utils.getEnv("EMBEDDING_DIMENSIONS", "1536"));
        String model = Utils.getEnv("AZURE_OPENAI_EMBEDDING_MODEL", "text-embedding-3-small");

        System.out.println("==============================================");
        System.out.println("  Azure DocumentDB - Compare All Algorithms");
        System.out.println("==============================================");
        System.out.printf("  Query:   \"%s\"%n", queryText);
        System.out.printf("  Top K:   %d%n", topK);
        System.out.printf("  Metrics: COS, L2, IP%n");
        System.out.printf("  Algos:   IVF, HNSW, DiskANN%n");
        System.out.println();

        List<SearchResult> results = new ArrayList<>();

        try (MongoClient mongoClient = Utils.getMongoClient()) {
            MongoDatabase database = mongoClient.getDatabase(databaseName);
            MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME);

            // Load data ONCE into the single collection
            System.out.println("  Loading data from: " + dataFile);
            List<Document> data = Utils.readJsonFile(dataFile);
            System.out.printf("  Loaded %d documents%n", data.size());

            collection.drop();
            System.out.println("  Collection reset.");
            Utils.insertData(collection, data, 100);

            // Generate ONE embedding for the query (reused for all 9 searches)
            OpenAIClient aiClient = Utils.getOpenAIClient();
            System.out.printf("%n  Generating embedding for: \"%s\"%n", queryText);
            List<Float> queryVector = Utils.getEmbedding(aiClient, queryText, model);
            System.out.printf("  Embedding generated (%d dimensions)%n%n", queryVector.size());

            // Convert to doubles for BSON
            List<Double> vectorAsDoubles = queryVector.stream()
                    .map(Float::doubleValue)
                    .toList();

            // Run 9 algorithm × metric combinations sequentially (create→search→drop)
            // DocumentDB does not allow multiple vector indexes of the same kind
            // on the same field path simultaneously.
            System.out.println("  Running 9 algorithm × metric combinations...\n");
            for (String algo : ALGORITHMS) {
                for (String metric : METRICS) {
                    String indexName = String.format("vector_%s_%s", algo, metric.toLowerCase());

                    // 1. Drop all existing vector indexes
                    dropVectorIndexes(collection, vectorField);

                    // 2. Create this specific index
                    createIndex(database, collection, vectorField, dimensions, algo, metric);
                    System.out.printf("  ✓ %s created%n", indexName);

                    // 3. Wait for index to build
                    try { Thread.sleep(5000); } catch (InterruptedException e) { Thread.currentThread().interrupt(); }

                    // 4. Search
                    List<Document> searchResults = performSearch(
                            collection, vectorAsDoubles, vectorField, topK);

                    // 5. Extract top 2 results
                    String top1Name = "-"; double top1Score = 0.0;
                    String top2Name = "-"; double top2Score = 0.0;
                    if (!searchResults.isEmpty()) {
                        Document top1 = searchResults.get(0);
                        top1Name = top1.getString("HotelName") != null ? top1.getString("HotelName") : "-";
                        top1Score = top1.getDouble("score") != null ? top1.getDouble("score") : 0.0;
                    }
                    if (searchResults.size() > 1) {
                        Document top2 = searchResults.get(1);
                        top2Name = top2.getString("HotelName") != null ? top2.getString("HotelName") : "-";
                        top2Score = top2.getDouble("score") != null ? top2.getDouble("score") : 0.0;
                    }
                    results.add(new SearchResult(algo.toUpperCase(), metric, top1Name, top1Score, top2Name, top2Score));
                }
            }

            // Cleanup: drop the comparison collection
            System.out.println("\n  Cleanup: dropping comparison collection...");
            collection.drop();
            System.out.println("  Cleanup: dropped collection 'hotels'");
        }

        // Print comparison table
        printComparisonTable(results);
    }

    private static void dropVectorIndexes(MongoCollection<Document> collection, String vectorField) {
        for (Document idx : collection.listIndexes()) {
            String name = idx.getString("name");
            Document key = idx.get("key", Document.class);
            if (key != null && "cosmosSearch".equals(key.getString(vectorField))) {
                try {
                    collection.dropIndex(name);
                } catch (Exception e) {
                    // Ignore if index doesn't exist
                }
            }
        }
    }

    private static void createIndex(MongoDatabase database, MongoCollection<Document> collection,
                                    String vectorField, int dimensions,
                                    String algo, String metric) {
        String indexName = String.format("vector_%s_%s", algo, metric.toLowerCase());

        Document cosmosSearchOptions = new Document()
                .append("dimensions", dimensions)
                .append("similarity", metric);

        switch (algo) {
            case "ivf" -> cosmosSearchOptions
                    .append("kind", "vector-ivf")
                    .append("numLists", 1);
            case "hnsw" -> cosmosSearchOptions
                    .append("kind", "vector-hnsw")
                    .append("m", 16)
                    .append("efConstruction", 64);
            case "diskann" -> cosmosSearchOptions
                    .append("kind", "vector-diskann")
                    .append("maxDegree", 32)
                    .append("lBuild", 50);
        }

        Document indexDefinition = new Document()
                .append("name", indexName)
                .append("key", new Document(vectorField, "cosmosSearch"))
                .append("cosmosSearchOptions", cosmosSearchOptions);

        Document command = new Document("createIndexes", collection.getNamespace().getCollectionName())
                .append("indexes", List.of(indexDefinition));

        try {
            database.runCommand(command);
        } catch (Exception e) {
            // Idempotent: ignore if index already exists
            if (!e.getMessage().contains("already exists")) {
                throw e;
            }
        }
    }

    private static List<Document> performSearch(MongoCollection<Document> collection,
                                                List<Double> vectorAsDoubles,
                                                String vectorField, int topK) {
        Document searchStage = new Document("$search", new Document("cosmosSearch", new Document()
                .append("vector", vectorAsDoubles)
                .append("path", vectorField)
                .append("k", topK)));

        Document projectStage = new Document("$project", new Document()
                .append("_id", 0)
                .append("HotelName", 1)
                .append("Description", 1)
                .append("score", new Document("$meta", "searchScore")));

        List<Document> pipeline = List.of(searchStage, projectStage);
        List<Document> results = new ArrayList<>();
        collection.aggregate(pipeline).forEach(results::add);
        return results;
    }

    private static void printComparisonTable(List<SearchResult> results) {
        System.out.println("┌──────────┬────────┬────────────────────────────┬────────┬────────────────────────────┬────────┬───────┐");
        System.out.printf("│ %-9s│ %-7s│ %-27s│ %-7s│ %-27s│ %-7s│ %-6s│%n",
                "Algorithm", "Metric", "Top 1 Result", "Score", "Top 2 Result", "Score", "Diff");
        System.out.println("├──────────┼────────┼────────────────────────────┼────────┼────────────────────────────┼────────┼───────┤");

        for (int i = 0; i < results.size(); i++) {
            SearchResult r = results.get(i);
            double diff = Math.abs(r.top1Score() - r.top2Score());
            String top1Display = r.top1Name().length() > 27 ? r.top1Name().substring(0, 24) + "..." : r.top1Name();
            String top2Display = r.top2Name().length() > 27 ? r.top2Name().substring(0, 24) + "..." : r.top2Name();
            System.out.printf("│ %-9s│ %-7s│ %-27s│ %-7.4f│ %-27s│ %-7.4f│ %-6.4f│%n",
                    r.algorithm(), r.metric(), top1Display, r.top1Score(), top2Display, r.top2Score(), diff);
            if (i < results.size() - 1) {
                System.out.println("├──────────┼────────┼────────────────────────────┼────────┼────────────────────────────┼────────┼───────┤");
            }
        }
        System.out.println("└──────────┴────────┴────────────────────────────┴────────┴────────────────────────────┴────────┴───────┘");
    }

    private record SearchResult(
            String algorithm,
            String metric,
            String top1Name,
            double top1Score,
            String top2Name,
            double top2Score) {
    }
}
